src/Workspaces/Core/Portable/FindSymbols/FindReferences/FindReferencesSearchEngine.UnidirectionalSymbolSet.cs

```diff
@@ -0,0 +1,69 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+
+namespace Microsoft.CodeAnalysis.FindSymbols
+{
+    internal partial class FindReferencesSearchEngine
+    {
+        /// <summary>
+        /// Symbol set used when <see cref="FindReferencesSearchOptions.UnidirectionalHierarchyCascade"/> is <see
+        /// langword="true"/>.  This symbol set will only cascade in a uniform direction once it walks either up or down
+        /// from the initial set of symbols. This is the symbol set used for features like 'Find Refs', where we only
+        /// want to return location results for members that could feasible actually end up calling into that member at
+        /// runtime.  See the docs of <see cref="FindReferencesSearchOptions.UnidirectionalHierarchyCascade"/> for more
+        /// information on this.
+        /// </summary>
+        private sealed class UnidirectionalSymbolSet : SymbolSet
+        {
+            private readonly HashSet<ISymbol> _initialAndDownSymbols;
+
+            /// <summary>
+            /// When we're doing a unidirectional find-references, the initial set of up-symbols can never change.
+            /// That's because we have computed the up set entirely up front, and no down symbols can produce new
+            /// up-symbols (as going down then up would not be unidirectional).
+            /// </summary>
+            private readonly ImmutableHashSet<ISymbol> _upSymbols;
+
+            public UnidirectionalSymbolSet(FindReferencesSearchEngine engine, HashSet<ISymbol> initialSymbols, HashSet<ISymbol> upSymbols)
+                : base(engine)
+            {
+                _initialAndDownSymbols = initialSymbols;
+                _upSymbols = upSymbols.ToImmutableHashSet();
+            }
+
+            public override ImmutableArray<ISymbol> GetAllSymbols()
+            {
+                using var _ = ArrayBuilder<ISymbol>.GetInstance(out var result);
```

Comment:
```suggestion
                using var _ = ArrayBuilder<ISymbol>.GetInstance(_upSymbols.Count + _initialAndDownSymbols.Count, out var result);
```

I'm not 100% sure that GetInstance exists and can take an initial size, but it seems likely. This also assumes it's optimal to ask for the correct size upfront

Reply:
done.